src/ch/ch_interface.c

```diff
@@ -0,0 +1,126 @@
+#include "ch_interface.h"
+#include "domain_nwfilter.h"
+#include "virnetdev.h"
+#include "virnetdevtap.h"
+#include "network_conf.h"
+#include "virlog.h"
+#include "viralloc.h"
+#include "domain_audit.h"
+#include "virnetdevbridge.h"
+#include "virebtables.h"
+#include "virfile.h"
+#include <fcntl.h>
+
+#define VIR_FROM_THIS VIR_FROM_CH
+
+VIR_LOG_INIT("ch.ch_interface");
+
+/* chInterfaceBridgeConnect:
+ * @def: the definition of the VM
+ * @driver: qemu driver data
+ * @net: pointer to the VM's interface description
+ * @tapfd: array of file descriptor return value for the new device
+ * @tapfdsize: number of file descriptors in @tapfd
+ *
+ * Called *only* called if actualType is VIR_DOMAIN_NET_TYPE_NETWORK or
+ * VIR_DOMAIN_NET_TYPE_BRIDGE (i.e. if the connection is made with a tap
+ * device connecting to a bridge device)
+ */
+int
+chInterfaceBridgeConnect(virDomainDefPtr def,
+                           virCHDriverPtr driver,
+                           virDomainNetDefPtr net,
+                           int *tapfd,
+                           size_t *tapfdSize)
+{
+    const char *brname;
+    int ret = 0;
+    unsigned int tap_create_flags = VIR_NETDEV_TAP_CREATE_IFUP;
+    bool template_ifname = false;
+    virCHDriverConfigPtr cfg = virCHDriverGetConfig(driver);
+    const char *tunpath = "/dev/net/tun";
+
+    if (net->backend.tap) {
+        tunpath = net->backend.tap;
+        if (!driver->privileged) {
+            virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                           _("cannot use custom tap device in session mode"));
+            goto cleanup;
+        }
+    }
+
+    if (!(brname = virDomainNetGetActualBridgeName(net))) {
+        virReportError(VIR_ERR_INTERNAL_ERROR, "%s", _("Missing bridge name"));
+        goto cleanup;
+    }
+
+    if (!net->ifname ||
+        STRPREFIX(net->ifname, VIR_NET_GENERATED_TAP_PREFIX) ||
+        strchr(net->ifname, '%')) {
+        VIR_FREE(net->ifname);
```

Comment:
I keep forgetting when I grab the qemu code but the VIR_ memory management is going away so I recommend using the glib [functions](https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html#g-free) so we don't need to fix this later.